File: 684. Redundant Connection/findRedundantConnection.py
from typing import List

class UnionFind:
  def __init__(self, n: int) -> None:
    self.id = [i for i in range(n)]
    self.rank = [0] * n

  def union(self, x: int, y: int) -> None:
    rootX = self.find(x)
    rootY = self.find(y)

    if rootX == rootY:
      return False

    if self.rank[rootX] < self.rank[rootY]:
      self.id[rootX] = rootY
    elif self.rank[rootX] > self.rank[rootY]:
      self.id[rootY] = rootX
    else:
      self.id[rootX] = rootY
      self.rank[rootY] += 1
    
    return True

  def find(self, x: int) -> int:
    if self.id[x] != x:
      self.id[x] = self.find(self.id[x])
    return self.id[x]
  
class Solution:
  def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
    uf = UnionFind(len(edges) + 1)

    for u, v in  edges:
      if not uf.union(u, v):
        return [u, v]
      

edges = [[1,2],[1,3],[2,3]]
solution = Solution()
print('findRedundantConnection: ', solution.findRedundantConnection(edges))

File: 684. Redundant Connection/findRedundantConnection.js
class UnionFind {
  constructor(n) {
    this.id = [];
    for (let i = 0; i < n; i++) {
      this.id[i] = i;
    }
    this.rank = Array(n).fill(0);
  }

  union(u, v) {
    const rootU = this.find(u);
    const rootV = this.find(v);

    if (rootU === rootV) {
      return false;
    }

    if(this.rank[rootU] < this.rank[rootV]) {
      this.id[rootU] = rootV;
    } else if(this.rank[rootU] > this.rank[rootV]) {
      this.id[rootV] = rootU;
    } else {
      this.id[rootU] = rootV;
      this.rank[rootV]++;
    }

    return true;
  }

  find(u) {
    if (this.id[u] !== u) {
      this.id[u] = this.find(this.id[u]);
    }

    return this.id[u];
  }
}
/**
 * @param {number[][]} edges
 * @return {number[]}
 */
const findRedundantConnection = (edges) => {
  const unionFind = new UnionFind(edges.length + 1);
  for (let edge of edges) {
    const [u, v] = edge;
    if (!unionFind.union(u, v)) {
      return edge;
    }
  }
};


const edges = [[1,2], [1,3], [2,3]];
console.log('findRedundantConnection: ', findRedundantConnection(edges)); // [2,3]
